lib/upipe/uclock_ptp.c

```diff
@@ -0,0 +1,222 @@
+ *
+ * @param urefcount pointer to urefcount
+ */
+static void uclock_ptp_free(struct urefcount *urefcount)
+{
+    struct uclock_ptp *ptp = uclock_ptp_from_urefcount(urefcount);
+    urefcount_clean(urefcount);
+    for (int i = 0; i < 2; i++) {
+        ubase_clean_fd(&ptp->fd[i]);
+#ifdef __linux__
+        ubase_clean_fd(&ptp->if_fd[i]);
+#endif
+    }
+    free(ptp);
+}
+
+/** @internal */
+static int uclock_ptp_nic_clock_idx(struct uclock_ptp *ptp,
+        struct uprobe *uprobe, int i, const char *interface)
+{
+#ifdef __linux__
+    struct ethtool_ts_info info;
+
+    memset(&info, 0, sizeof(info));
+    info.cmd = ETHTOOL_GET_TS_INFO;
+    strncpy(ptp->ifr[i].ifr_name, interface, IFNAMSIZ - 1);
+    ptp->ifr[i].ifr_data = (char *) &info;
+
+    if (ioctl(ptp->if_fd[i], SIOCETHTOOL, &ptp->ifr[i]) < 0) {
+        uprobe_err_va(uprobe, NULL, "Couldn't get ethtool ts information for %s: %m",
+            interface);
+        info.phc_index = -1;
+    }
+
+    return info.phc_index;
+#else
+    /* TODO */
+    return -1;
+#endif
+}
+
+/** @internal */
+static int uclock_ptp_open_nic(struct uclock_ptp *ptp, struct uprobe *uprobe,
+        int i, const char *interface)
+{
+    int idx = uclock_ptp_nic_clock_idx(ptp, uprobe, i, interface);
+
+    if (idx < 0) {
+        uprobe_err_va(uprobe, NULL, "No PTP device found for %s", interface);
+        return UBASE_ERR_EXTERNAL;
+    }
+
+    char clkdev[32];
+    snprintf(clkdev, sizeof(clkdev), "/dev/ptp%u", idx);
+
+    ptp->fd[i] = open(clkdev, O_RDWR);
+    if (ptp->fd[i] < 0) {
+        uprobe_err_va(uprobe, NULL, "Could not open PTP device %s: %m", clkdev);
+        return UBASE_ERR_EXTERNAL;
+    }
+
+    return UBASE_ERR_NONE;
+}
+
+/** @This allocates a new uclock structure.
+ *
+ * @param uprobe probe catching log events for error reporting
+ * @param interface NIC names
+ * @return pointer to uclock, or NULL in case of error
+ */
+struct uclock *uclock_ptp_alloc(struct uprobe *uprobe, const char *interface[2])
+{
+    struct uclock_ptp *ptp = malloc(sizeof(struct uclock_ptp));
+    if (unlikely(ptp == NULL))
+        return NULL;
+
+    struct urefcount *urefcount = uclock_ptp_to_urefcount(ptp);
+
+    urefcount_init(urefcount, uclock_ptp_free);
+    ptp->uclock.refcount = urefcount;
+    ptp->uclock.uclock_now = uclock_ptp_now;
+    ptp->uclock.uclock_to_real = NULL;
+    ptp->uclock.uclock_from_real = NULL;
+
+    for (int i = 0; i < 2; i++) {
+        ptp->fd[i] = -1;
+#ifdef __linux__
+        ptp->if_fd[i] = -1;
+#endif
+    }
+
+    for (int i = 0; i < 2 && interface[i]; i++) {
+#ifdef __linux__
+        ptp->if_fd[i] = socket(AF_INET, SOCK_DGRAM, 0);
+        if (ptp->if_fd[i] < 0) {
+            uprobe_err_va(uprobe, NULL, "%s: can't open socket (%m)",
+                    interface[i]);
+            goto err;
+        }
+        memset(&ptp->ifr[i], 0, sizeof(ptp->ifr[i]));
+#endif
+        if (!ubase_check(uclock_ptp_open_nic(ptp, uprobe, 0, interface[i])))
```

Comment:
95% sure 0 should be replaced with i here

Reply:
Seems so